### PR TITLE
Add EIP-6206: EOF - JUMPF instruction

### PR DESCRIPTION
**Pull request by lightclient**
This is a follow up from #6205 which removed `JUMPF` specification from EIP-4750. We have reintroduced it with a direct dependency to EIP-5450 to avoid specifying any runtime validation semantics (which would ultimately be discarded with EIP-5450).

### CODE REVIEW

File: EIPS/eip-6206.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+4. Pops nothing and pushes nothing to data stack.
+
+### Code Validation
+
+Let the definition of `type[i]` be inherited from [EIP-4750](./eip-4750.md) and define `stack_height` to be the height of the stack at a certain instruction during the instruction flow traversal if the data stack at the start of the function were equal to `type[i].inputs`.
+
+* Immediate argument of `JUMPF` MUST be greater than or equal to the total number of code sections.
+* Stack height at `JUMPF` MUST be greater than or equal to `type[code_section_index].inputs`.
+* `type[current_section_index].outputs` MUST equal `stack_height - type[code_section_index].inputs + type[code_section_index].outputs`. This means that `code_section_index` can output less stack elements than the original code section called by the top element on the return stack if the `current_section_index` code section leaves the delta `type[current_section_index] - type[code_section_index]` element(s) on the stack.
+* Extend [EIP-4200](./eip-4200) code validation to fail if any `RJUMP*` offset points to one of the two bytes directly following a `JUMPF` instruction.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md]: @SamWilsn, why did EIPW not catch this previously incorrect link? (was `[EIP-4200](./eip-4200)`)

File: EIPS/eip-6206.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+status: Draft
+type: Standards Track
+category: Core
+created: 2022-12-21
+requires: 4750, 5450
+---
+
+## Abstract
+
+Allow tail call optimizations in [EOF functions](./eip-4750.md) by introducing a new instruction `JUMPF`, which jumps to a code sections without adding a new return stack frame.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:16]:
```suggestion
This EIP allows for tail call optimizations in EOF functions ([EIP-4750](./eip-4750.md)) by introducing a new instruction `JUMPF`, which jumps to a code section without adding a new return stack frame.
```
Again, eipw didn't catch this somehow

File: EIPS/eip-6206.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+## Motivation
+
+It is common for function to make a call at the end of the routine only to then return. `JUMPF` optimizes this behavior by changing code sections without needing to update the return stack.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:20]:
```suggestion
It is common for functions to make a call at the end of the routine only to then return. `JUMPF` optimizes this behavior by changing code sections without needing to update the return stack.
```

File: EIPS/eip-6206.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+## Specification
+
+a new instruction `JUMPF (0xb2)` is introduced.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:24]:
```suggestion
A new instruction, `JUMPF (0xb2)`, is introduced.
```

File: EIPS/eip-6206.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+### Execution Semantics
+
+1. Has one immediate argument, `code_section_index`, encoded as a 16-bit unsigned big-endian value.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:28]:
```suggestion
1. `JUMPF` has one immediate argument, `code_section_index`, encoded as a 16-bit unsigned big-endian value.
```

File: EIPS/eip-6206.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+2. If data stack size exceeds `1024 - type[code_section_index].max_stack_height` (i.e. if the called function may exceed the global stack height limit), execution results in exceptional halt. This also guarantees that the stack height after the call is within the limits.
+3. Charges 5 gas.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:30]:
```suggestion
3. `JUMPF` costs 5 gas.
```

File: EIPS/eip-6206.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+It is common for function to make a call at the end of the routine only to then return. `JUMPF` optimizes this behavior by changing code sections without needing to update the return stack.
+
+## Specification
+
+a new instruction `JUMPF (0xb2)` is introduced.
+
+### Execution Semantics
+
+1. Has one immediate argument, `code_section_index`, encoded as a 16-bit unsigned big-endian value.
+2. If data stack size exceeds `1024 - type[code_section_index].max_stack_height` (i.e. if the called function may exceed the global stack height limit), execution results in exceptional halt. This also guarantees that the stack height after the call is within the limits.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:29]:
```suggestion
2. If the data stack size exceeds `1024 - type[code_section_index].max_stack_height` (i.e. if the called function may exceed the global stack height limit), execution results in an exceptional halt. This guarantees that the stack height after the call is within the limits.
```

File: EIPS/eip-6206.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+3. Charges 5 gas.
+4. Pops nothing and pushes nothing to data stack.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:31]:
```suggestion
4. `JUMPF` neither pops nor pushes anything to the data stack.
```

File: EIPS/eip-6206.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+### Code Validation
+
+Let the definition of `type[i]` be inherited from [EIP-4750](./eip-4750.md) and define `stack_height` to be the height of the stack at a certain instruction during the instruction flow traversal if the data stack at the start of the function were equal to `type[i].inputs`.
+
+* Immediate argument of `JUMPF` MUST be greater than or equal to the total number of code sections.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:37]:
```suggestion
* The immediate argument of `JUMPF` MUST be greater than or equal to the total number of code sections.
```

File: EIPS/eip-6206.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+* Stack height at `JUMPF` MUST be greater than or equal to `type[code_section_index].inputs`.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:38]:
```suggestion
* The stack height at `JUMPF` MUST be greater than or equal to `type[code_section_index].inputs`.
```

File: EIPS/eip-6206.md [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+* `type[current_section_index].outputs` MUST equal `stack_height - type[code_section_index].inputs + type[code_section_index].outputs`. This means that `code_section_index` can output less stack elements than the original code section called by the top element on the return stack if the `current_section_index` code section leaves the delta `type[current_section_index] - type[code_section_index]` element(s) on the stack.
+* Extend [EIP-4200](./eip-4200.md) code validation to fail if any `RJUMP*` offset points to one of the two bytes directly following a `JUMPF` instruction.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:40]:
```suggestion
* The code validation defined in [EIP-4200](./eip-4200.md) also fails if any `RJUMP*` offset points to one of the two bytes directly following a `JUMPF` instruction.
```

File: EIPS/eip-6206.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+As long as `JUMPF` prepares the delta `type[current_section_index] - type[code_section_index]` stack elements before changing code sections, it is possible to jump to a section with less outputs than was originally entered via `CALLF`. This will reduce duplicated code as it will allow compilers more flexibility during code generation such that certain helpers can be used generically by functions, regardless of their output values.
+
+## Backwards Compatibility
+
+This change is backward compatible as EOF does not allow undefined instructions to be used or deployed, meaning no contracts will be affected.
+
+## Security Considerations
+
+None.
```
Comment (Pandapip1) [inline EIPS/eip-6206.md:54]:
```suggestion
Needs discussion.
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

**Review by Pandapip1 — APPROVED** [review]
LGTM now

**Comment by lightclient** [comment]
@Pandapip1 I don't think you should be be committing to other people's EIPs without their consent. Some of your changes are good, but as an author I would like to incorporate your feedback how I see fit.

**Comment by Pandapip1** [comment]
There are some that I will commit w/o consent (typo fixes, formatting errors, and the like). The only real one that I could see anybody taking issue with is https://github.com/ethereum/EIPs/pull/6206#discussion_r1055731295, which I was on the fence about including in the commit. In the future, I'll limit it a bit further. Thanks for the feedback!

**Pull request merged**